fleet_adapter_mir_actions/fleet_adapter_mir_actions/rmf_cart_delivery.py

```diff
@@ -0,0 +1,512 @@
+import enum
+import numpy as np
+from typing import Any
+from dataclasses import dataclass
+import importlib
+import requests
+from urllib.error import HTTPError
+from fleet_adapter_mir_actions.mir_action import MirAction, MirActionFactory
+from fleet_adapter_mir.mir_api import MirAPI, MirStatus, MiRStateCode
+
+
+class PickupState(enum.IntEnum):
+    PICKUP_INITIALIZED = 0
+    PICKUP_ALLOCATED = 1
+    MOVE_REQUESTED = 2
+    AT_PICKUP = 3
+    DOCK_REQUESTED = 4
+    DOCK_COMPLETED = 5
+    PICKUP_REQUESTED = 6
+    PICKUP_SUCCESS = 7
+    TASK_CANCELLED = 8
+
+
+class Mission:
+    def __init__(self, queue_id: str, start_time: float):
+        self.queue_id = queue_id
+        self.start_time = start_time
+
+
+@dataclass
+class Pickup:
+    state: PickupState
+    pickup_lots: list[str]  # contains list of pickup lots
+    cart_id: str
+    execution: Any
+    mission: Mission
+    latching: bool
+
+
+@dataclass
+class Dropoff:
+    execution: Any
+    mission: Mission
+
+
+class ActionFactory(MirActionFactory):
+    def make_action(self,
+                    node,
+                    name,
+                    mir_api,
+                    update_handle,
+                    fleet_config,
+                    action_config) -> MirAction:
+        return CartDelivery(
+            node, name, mir_api, update_handle, fleet_config, action_config)
+
+
+class CartDelivery(MirAction):
+    def __init__(
+            self,
+            node,
+            name,
+            mir_api,
+            update_handle,
+            fleet_config,
+            action_config):
+        MirAction.__init__(
+            self, node, name, mir_api, update_handle, fleet_config,
+            action_config)
+
+        self.pickup: Pickup = None
+        self.dropoff: Dropoff = None
+        self.search_timeout = \
+            self.action_config.get('search_timeout', 60)  # seconds
+
+        # Store the mission names to be used during the action
+        self.dock_to_cart_mission = \
+            self.action_config['missions']['dock_to_cart']
+        self.pickup_mission = \
+            self.action_config['missions']['pickup']
+        self.dropoff_mission = \
+            self.action_config['missions']['dropoff']
+        self.exit_mission = \
+            self.action_config['missions']['exit_lot']
+        self.footprint_mission = \
+            self.action_config['missions'].get('update_footprint')  # Optional
+
+        # Initialize cart marker type
+        self.cart_marker_type_guid = \
+            self.api.docking_offsets_guid_get(
+                self.action_config['marker_types']['cart'])
+
+        # Import CartDetection module if provided
+        detection_module = self.action_config.get('cart_detection_module')
+        assert (detection_module is not None,
+                'CartDetection module is required for CartDelivery plugin, ' +
+                'but it is not found!')
+        detection_plugin = importlib.import_module(detection_module)
```

Comment:
Ideally this should be something parsed and loaded by the `ActionFactory` during the configuration step while the fleet adapter is starting up.